R/pr.R

```diff
@@ -0,0 +1,436 @@
+#' also invisibly returns the updated router.
+#'
+#' @param pr the host plumber router.
+#' @param path a character string. Where to mount router.
+#' @param router a plumber router. Router to be mounted.
+#'
+#' @return a plumber router with the supplied router mounted
+#'
+#' @examples
+#' \dontrun{
+#' pr1 <- pr_new() %>%
+#'   pr_get("/hello", function() "Hello")
+#'
+#' pr_new() %>%
+#'   pr_get("/goodbye", function() "Goodbye") %>%
+#'   pr_mount("/hi", pr1) %>%
+#'   pr_run()
+#' }
+#'
+#' @export
+pr_mount <- function(pr,
+                     path,
+                     router) {
+  pr$mount(path = path, router = router)
+  invisible(pr)
+}
+
+#' Register a hook
+#'
+#' Plumber routers support the notion of "hooks" that can be registered
+#' to execute some code at a particular point in the lifecycle of a request.
+#' Plumber routers currently support four hooks:
+#' \enumerate{
+#'  \item `preroute(data, req, res)`
+#'  \item `postroute(data, req, res, value)`
+#'  \item `preserialize(data, req, res, value)`
+#'  \item `postserialize(data, req, res, value)`
+#' }
+#' In all of the above you have access to a disposable environment in the `data`
+#' parameter that is created as a temporary data store for each request. Hooks
+#' can store temporary data in these hooks that can be reused by other hooks
+#' processing this same request.
+#'
+#' One feature when defining hooks in Plumber routers is the ability to modify
+#' the returned value. The convention for such hooks is: any function that accepts
+#' a parameter named `value` is expected to return the new value. This could
+#' be an unmodified version of the value that was passed in, or it could be a
+#' mutated value. But in either case, if your hook accepts a parameter
+#' named `value`, whatever your hook returns will be used as the new value
+#' for the response.
+#'
+#' You can add hooks using the `pr_register_hook`, or you can add multiple
+#' hooks at once using `pr_register_hooks`, which takes a named list in
+#' which the names are the names of the hooks, and the values are the
+#' handlers themselves.
+#'
+#' @param pr a plumber router
+#' @param stage a character string. Point in the lifecycle of a request.
+#' @param handler a hook function.
+#'
+#' @return a plumber router with the defined hook(s) added
+#'
+#' @examples
+#' \dontrun{
+#' pr_new() %>%
+#'   pr_register_hook("preroute", function(req){
+#'     cat("Routing a request for", req$PATH_INFO, "...\n")
+#'   }) %>%
+#'   pr_register_hooks(list(
+#'   preserialize = function(req, value){
+#'     print("About to serialize this value:")
+#'     print(value)
+#'
+#'     # Must return the value since we took one in. Here we're not choosing
+#'     # to mutate it, but we could.
+#'     value
+#'   },
+#'   postserialize = function(res){
+#'     print("We serialized the value as:")
+#'     print(res$body)
+#'   }
+#'  )) %>%
+#'  pr_handle("GET", "/", function(){ 123 }) %>%
+#'  pr_run()
```

Comment:
```suggestion
#' pr_new() %>%
#'   pr_register_hook("preroute", function(req){
#'     cat("Routing a request for", req$PATH_INFO, "...\n")
#'   }) %>%
#'   pr_register_hooks(list(
#'     preserialize = function(req, value){
#'       print("About to serialize this value:")
#'       print(value)
#'
#'       # Must return the value since we took one in. Here we're not choosing
#'       # to mutate it, but we could.
#'       value
#'     },
#'     postserialize = function(res){
#'       print("We serialized the value as:")
#'       print(res$body)
#'     }
#'   )) %>%
#'   pr_handle("GET", "/", function(){ 123 }) %>%
#'   pr_run()
```